pkg/distributor/query.go

```diff
@@ -184,6 +212,82 @@ func (d *Distributor) queryIngesters(ctx context.Context, replicationSet ring.Re
 	return result, nil
 }
 
+// mergeExemplarSets merges and dedupes two sets of already sorted exemplar pairs.
+// Both a and b should be lists of exemplars from the same series.
+// Defined here instead of pkg/util to avoid a import cycle.
+func mergeExemplarSets(a, b []cortexpb.Exemplar) []cortexpb.Exemplar {
+	result := make([]cortexpb.Exemplar, 0, len(a)+len(b))
+	i, j := 0, 0
+	for i < len(a) && j < len(b) {
+		if a[i].TimestampMs < b[j].TimestampMs {
+			result = append(result, a[i])
+			i++
+		} else if a[i].TimestampMs > b[j].TimestampMs {
+			result = append(result, b[j])
+			j++
+		} else {
+			result = append(result, a[i])
+			i++
+			j++
+		}
+	}
+	// Add the rest of a or b. One of them is empty now.
+	result = append(result, a[i:]...)
+	result = append(result, b[j:]...)
+	return result
+}
+
+// queryIngestersExemplars queries the ingesters for exemplars.
+func (d *Distributor) queryIngestersExemplars(ctx context.Context, replicationSet ring.ReplicationSet, req *ingester_client.ExemplarQueryRequest) (*ingester_client.ExemplarQueryResponse, error) {
+	// Fetch exemplars from multiple ingesters in parallel, using the replicationSet
+	// to deal with consistency.
+	results, err := replicationSet.Do(ctx, d.cfg.ExtraQueryDelay, func(ctx context.Context, ing *ring.InstanceDesc) (interface{}, error) {
+		client, err := d.ingesterPool.GetClientFor(ing.Addr)
+		if err != nil {
+			return nil, err
+		}
+
+		resp, err := client.(ingester_client.IngesterClient).QueryExemplars(ctx, req)
+		d.ingesterQueries.WithLabelValues(ing.Addr).Inc()
+		if err != nil {
+			d.ingesterQueryFailures.WithLabelValues(ing.Addr).Inc()
+			return nil, err
+		}
+
+		return resp, nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Merge results from replication set.
+	var keys []string
+	exemplarResults := make(map[string]cortexpb.TimeSeries)
+	for _, result := range results {
+		r := result.(*ingester_client.QueryResponse)
+		for _, ts := range r.Timeseries {
+			lbls := cortexpb.FromLabelAdaptersToLabels(ts.Labels).String()
+			e, ok := exemplarResults[lbls]
+			if !ok {
+				exemplarResults[lbls] = ts
+				keys = append(keys, lbls)
+			}
+			// Merge in any missing values from another ingesters exemplars for this series.
+			e.Exemplars = mergeExemplarSets(e.Exemplars, ts.Exemplars)
```

Comment:
**suggestion:** I may be missing something, but if I'm not mistaken, there is no need to call this function if there isn't an existing result, correct?

```suggestion
			} else {
				// Merge in any missing values from another ingesters exemplars for this series.
				e.Exemplars = mergeExemplarSets(e.Exemplars, ts.Exemplars)
			}
```

Reply:
Yeah probably not, I don't think it's functionally any different. I guess there's the potential for slightly more memory being allocated than is really necessary without the else. Is this what you were thinking?